doc/compiler/il/notes_mainline_outOfLine_GlRegDeps.md

```diff
@@ -0,0 +1,97 @@
+
+SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+-->
+
+# Mainline and out of line code
+The mainline path handles the common operations in a straight line to avoid branching or using branch prediction resources. Arranging the commonly run code on the mainline also helps reserve instruction cache. The work can be done in the same cache line by falling through. There is no need to jump to another spot to bring in an entire new cache line to the cache.
+
+Executing the common code on the mainline is not only adopted in IL but also in the code gen, such as the non-trivial logics: write barrier, instanceof, etc. The instructions that are believed to run commonly are arranged on the mainline path. Branching is chosen for the less common code.
+
+One example in the code gen is the bound check. It does the comparison between arraylength and index. If it fails, an exception is going to be raised.  Then it will branch to an out of line, or a snippet of code that jumps into the VM to throw the exception that needs to be thrown. Because the path that jumps to throw the exception is expected to be very rarely called, the snippet code is put at the end of the code section, or at the end of the generated code for that method, aka out of line, so that it does not interfere with the code that we believe will actually run.
+
+Another example is block ordering and extension of basic blocks in IL. Both uses the same logic to arrange the blocks. For example there is a block ending with `if`. One path is marked as cold. Another path is not marked as cold. Block ordering takes the cold path out of the mainline, puts it at the end of the method, and keeps the non-cold path as the fall through. This way it avoids consuming branch prediction and instruction cache resources.
+
+
+# Global register dependency
+
+`GlRegDeps` at its core is really a code gen register dependency. It associates virtual registers with real registers, or pairs virtual registers with real registers.
+
+Register dependencies enforce constraints. It says that this value has to be in this register at this time. Global register dependency is fundamentally about conveying particular values in particular registers across the block boundaries. The values and registers are expressed by `GlRegDeps` from one basic block to another.
+
+Most common control flow opcodes are: `if`, `goto`, `return`, `athrow`, `switch`. They should be the last tree before the `BBEnd`. They should have their own `GlRegDeps` for the branch target if required. The `GlRegDeps` under `BBEnd` is for the fall through case.
+
+If a value that is expected to be in a register is not in the register, or the register is not preserved if it is a call, it leads to register shuffling. The local register allocator has to shuffle it to a different register to make sure it has the right value in `GlRegDeps`. Register shuffling could come with a cost.
+
```

Comment:
You probably mean "make sure it has the right value in the right real register at the point where the `GlRegDeps` is evaluated".

You probably mean "Register shuffling does come at a cost" since there is no doubt that there is extra instruction path length to do the shuffling.